flang/lib/Lower/OpenACC.cpp

```diff
@@ -119,160 +119,233 @@ static Op createSimpleOp(Fortran::lower::FirOpBuilder &builder,
   return op;
 }
 
-static void genACC(Fortran::lower::AbstractConverter &converter,
-                   Fortran::lower::pft::Evaluation &eval,
-                   const Fortran::parser::OpenACCLoopConstruct &loopConstruct) {
-
-  const auto &beginLoopDirective =
-      std::get<Fortran::parser::AccBeginLoopDirective>(loopConstruct.t);
-  const auto &loopDirective =
-      std::get<Fortran::parser::AccLoopDirective>(beginLoopDirective.t);
+static void genAsyncClause(Fortran::lower::AbstractConverter &converter,
+                           const Fortran::parser::AccClause::Async *asyncClause,
+                           mlir::Value &async, bool &addAsyncAttr) {
+  const auto &asyncClauseValue = asyncClause->v;
+  if (asyncClauseValue) { // async has a value.
+    async = fir::getBase(converter.genExprValue(
+        *Fortran::semantics::GetExpr(*asyncClauseValue)));
+  } else {
+    addAsyncAttr = true;
+  }
+}
 
-  if (loopDirective.v == llvm::acc::ACCD_loop) {
+static void genDeviceTypeClause(
+    Fortran::lower::AbstractConverter &converter,
+    const Fortran::parser::AccClause::DeviceType *deviceTypeClause,
+    SmallVectorImpl<mlir::Value> &operands) {
+  const auto &deviceTypeValue = deviceTypeClause->v;
+  if (deviceTypeValue) {
+    for (const auto &scalarIntExpr : *deviceTypeValue) {
+      mlir::Value expr = fir::getBase(
+          converter.genExprValue(*Fortran::semantics::GetExpr(scalarIntExpr)));
+      operands.push_back(expr);
+    }
+  } else {
     auto &firOpBuilder = converter.getFirOpBuilder();
-    auto currentLocation = converter.getCurrentLocation();
+    // * was passed as value and will be represented as a -1 constant
+    // integer.
+    mlir::Value star = firOpBuilder.createIntegerConstant(
+        converter.getCurrentLocation(), firOpBuilder.getIntegerType(32),
+        /* STAR */ -1);
```

Comment:
I'd recommend having a `constexpr` function (it can return the -1 encoding) and use that instead of harder to find magic values.

Reply:
Sure I'll make the change.